### PR TITLE
Create guide for writing custom distributions

### PR DESCRIPTION
**Pull request by ally-lee**
## Description

Addresses issue #184 and aims to advance it to

### CODE REVIEW

File: examples/custom_distribution.ipynb
Comment (michaelosthege) [inline examples/custom_distribution.ipynb:1]: <p>And tests?</p><p> </p><p>And maybe to consider contributing it with a PR ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (michaelosthege) [inline examples/custom_distribution.ipynb:1]: <p>The next cell (implementation) should be introduced by the text. For example, that it should be an implementation of the above formula using just Aesara operations.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (michaelosthege) [inline examples/custom_distribution.ipynb:1]: <p>Can we use generator from, for example scipy?</p><p>(Yes we can, but maybe that should be mentioned explicitly.)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (michaelosthege) [inline examples/custom_distribution.ipynb:1]: <p>With https://github.com/pymc-devs/pymc3/pull/4867 we're adding a mechanism for setting initial values. The code example doesn't need to include that, but maybe it should be mentioned somewhere.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (michaelosthege) [inline examples/custom_distribution.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #14.            return generate_samples(genpoisson_rvs,
</pre><p>The code formatting should already be compatible with our pre-commit. In this case that'll be a line break after <code>generate_samples(</code>  followed by 4-space indentation.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (michaelosthege) [inline examples/custom_distribution.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.                          chains=2, cores=1, init='adapt_diag', random_seed=42)
</pre><p>For <code>v4</code>  the typical variable name would be <code>idata</code>  for the returned <code>InferenceData</code> object.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (michaelosthege) [inline examples/custom_distribution.ipynb:1]: <p><code>pm.traceplot</code>  is deprecated in favor of either <code>arviz.plot_trace</code>  or its alias <code>pm.plot_trace</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (michaelosthege) [inline examples/custom_distribution.ipynb:1]: <p>In <code>v4</code>  the <code>testval</code>  kwarg is deprecated in favor of <code>initval</code>.  Is it really necessary though? If so why?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

**Comment by michaelosthege** [comment]
@ally-lee nice work! I left some comments of which some may be nitpicky.

File: examples/custom_distribution.ipynb
Comment (OriolAbril) [reply examples/custom_distribution.ipynb:1]: <p>I find confusing that it uses <code>np.log</code>  and <code>tt.switch</code>  <span class="ql-cursor">﻿</span></p>

Comment (OriolAbril) [reply examples/custom_distribution.ipynb:1]: <p>Is this guide supposed to be for 3.x or for 4.x as of now?</p>

Comment (OriolAbril) [reply examples/custom_distribution.ipynb:1]: <p>even if using v3, all notebooks should already use <code>return_inferencedata=True</code>,  and if possible use coords and dims instead of shape.</p>

Comment (OriolAbril) [inline examples/custom_distribution.ipynb:1]: <p>Shouldn't "log likelihood" be "log probability". Distributions could technically be used for either prior or likelihood terms right?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (OriolAbril) [inline examples/custom_distribution.ipynb:1]: <p>Thinking in terms of https://github.com/pymc-devs/pymc3/issues/4899 (so it doesn't need to be done in this PR, we can add this to the issue and nothing else). pmf/pdf maybe also over/underdispersion look like terms that should go in the glossary.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (OriolAbril) [inline examples/custom_distribution.ipynb:1]: <p>is the data public? Can it be added to the <code>data</code>  folder and read with the try except described at https://github.com/pymc-devs/pymc3/wiki/PyMC3-Jupyter-Notebook-Style-Guide. I also noticed the notebook is on <code>examples/</code>  folder, I think it should go on <code>examples/pymc3_howtos</code>  one (as well as eventually being added to the table of contents so it appears on the web)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

**Comment by OriolAbril** [comment]
This looks great, sorry it took so long to review. I have added some comments related to latest updates and best practices with pymc3 and arviz which will require some changes to the plotting. I am happy to help with those changes

File: examples/custom_distribution.ipynb
Comment (ally-lee) [reply examples/custom_distribution.ipynb:1]: I don't think the `testval` is really serving any purpose. I think I just included it because it wasn't working without it.

Comment (ally-lee) [reply examples/custom_distribution.ipynb:1]: Yes this data is sourced from https://www.mass.gov/info-details/archive-of-covid-19-cases-in-massachusetts. I'll add it the csv I'm using to the `data` folder and move this file to `examples/pymc3_howtos`

Comment (ally-lee) [reply examples/custom_distribution.ipynb:1]: Good call. I kind of mention it with this sentence
```
It is meant to be analogous to `scipy.stats.<DIST_NAME>.rvs`.
```
but I should be more explicit about how it's not necessary to hardcode a generator if it already exists in `scipy`.

Comment (ally-lee) [reply examples/custom_distribution.ipynb:1]: Yes true good call

Comment (ally-lee) [reply examples/custom_distribution.ipynb:1]: Would it be helpful if I added these terms/definitions in a comment on that issue?

Comment (ally-lee) [reply examples/custom_distribution.ipynb:1]: I'll also follow the `try...except` clause recommended in the style guide

Comment (ally-lee) [reply examples/custom_distribution.ipynb]: I believe I'm still working with 3.x

**Comment by ally-lee** [comment]
@michaelosthege @OriolAbril Thank you both for your reviews! I made a bunch of updates based on your comments. Do you mind taking it over to make the more technical changes? I think that would be best since you're the experts on this technology. I'm feeling good about what I've submitted so far and am okay with whatever changes you need to make

File: examples/pymc3_howto/custom_distribution.ipynb
Comment (ricardoV94) [inline examples/pymc3_howto/custom_distribution.ipynb:1]: <p>There are two extra details that are worth mentioning:</p><p> </p><ol><li>For continuous distributions you also have to define the default transform, or inherit from a more specific class like <code>PositiveContinuous</code>  which specifies what the default transform should be. </li><li>On V3 you are supposed to also specify at least&nbsp;one ”default value" for the distribution during <code>init</code>  such as <code>self.mode</code>,  <code>self.median</code>  and <code>self.mean</code>  (the latter only for continuous distributions). This is used by some samplers or other compound distributions</li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (ricardoV94) [inline examples/pymc3_howto/custom_distribution.ipynb:1]: <p>I would suggest adding a prior predictive step, not only because it's part of the standard Bayesian worflow, but also because it uses your distribution in a different way than in <code>sample</code> or <code>posterior_predictive</code>, as an unobserved random variable. That in itself of a good sanity check that the distribution is well implemented and can be sampled from (with mcmc that is)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/185/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (ally-lee) [reply examples/pymc3_howto/custom_distribution.ipynb:1]: I'm trying to add this, but I think it won't work because [pm.AR](https://docs.pymc.io/api/distributions/timeseries.html#pymc3.distributions.timeseries.AR) doesn't have a `random` method. Is it ok if I leave this step out?

Comment (ally-lee) [reply examples/pymc3_howto/custom_distribution.ipynb:1]: Thank you for pointing these out! I'll add these notes to the text block. I also realized the reason I needed to set a `testval` when sampling was because I hadn't specified a "default value" for the distribution, so I can take that out now.

File: examples/custom_distribution.ipynb
Comment (ally-lee) [reply examples/custom_distribution.ipynb]: It turns out I needed the set a `testval` because I hadn't specified `self.mode` in the distribution. I'll remove it now that I'm specifying `self.mode`

File: examples/pymc3_howto/custom_distribution.ipynb
Comment (ricardoV94) [reply examples/pymc3_howto/custom_distribution.ipynb:1]: Actually what I had in mind was prior predictive sampling with `pm.sample` by simply removing the `observed` keyword. 

But that's not necessarily useful for this example...

Feel free to ignore it then

Comment (ally-lee) [reply examples/pymc3_howto/custom_distribution.ipynb:1]: I like your idea of doing a sanity check though! I think it would be nice to compare what samples look like from the standard Poisson vs generalized Poisson

Comment (ricardoV94) [reply examples/pymc3_howto/custom_distribution.ipynb:1]: <p>That's a good point!</p>

File: examples/custom_distribution.ipynb
Comment (OriolAbril) [reply examples/custom_distribution.ipynb]: That would be great yes! Thanks!

Comment (ally-lee) [reply examples/custom_distribution.ipynb]: Done!

**Comment by ally-lee** [comment]
Hey @OriolAbril just wanted to check in, are you ok to make those final changes related to latest updates/best practices? Is there anything else I can do to help get this ready to release?

**Comment by chiral-carbon** [comment]
hey @ally-lee, I would be up for making the final changes to get this notebook to best practices, if that's alright with you?
@MarcoGorelli would that be okay?

**Comment by MarcoGorelli** [comment]
sure, no objections from me

**Comment by chiral-carbon** [comment]
@MarcoGorelli would you merge this notebook then, if everything else (except using `return_inferencedata=True` in pm.sample and plotting using ArviZ) seems alright? because then I could open a new PR with the new changes.

**Comment by MarcoGorelli** [comment]
> @MarcoGorelli would you merge this notebook then, if everything else (except using `return_inferencedata=True` in pm.sample and plotting using ArviZ) seems alright? because then I could open a new PR with the new changes.

I haven't had a chance to look at this yet, but you could always make a new branch from this one so as to keep @ally-lee 's commits and to add your ones on top

**Comment by OriolAbril** [comment]
Then it's probably easier for you to open the PR to this branch/PR I think 🤔

**Comment by chiral-carbon** [comment]
either is fine tbh. I would also have doubts and probably make some changes incrementally, so if having lots of comments under this one PR would not be too messy then I'll just push here. otherwise, I can open a new PR, no issues there.

**Comment by chiral-carbon** [comment]
hey @ally-lee I tried pushing commits to this PR but I get a git error saying permission denied. could you check if the option "Allow edits from maintainers" is checked in the PR? it should be a checkbox on the right side of this page!

**Comment by ally-lee** [comment]
Hmm it's already checked
Screen Shot 2021-08-31 at 4 22 11 PM

**Comment by chiral-carbon** [comment]
oh okay :thinking: not sure what's wrong then

**Comment by MarcoGorelli** [comment]
> oh okay 🤔 not sure what's wrong then

can you show the command you ran, and the output?

**Comment by OriolAbril** [comment]
I think the problem is that @chiral-carbon only has triage permissions so she isn't seen as "maintainer" by github

**Comment by chiral-carbon** [comment]
@OriolAbril @MarcoGorelli I searched online and I think only the owners/maintainers of a project can push commits to a PR opened by someone else in the project. this is the error message I get on doing `git push allylee HEAD:custom-distribution` ( I had added @ally-lee's fork as a remote locally with `git remote add allylee https://github.com/ally-lee/pymc-examples.git`):
```
remote: Permission to ally-lee/pymc-examples.git denied to chiral-carbon.
fatal: unable to access 'https://github.com/ally-lee/pymc-examples.git/': The requested URL returned error: 403
```
will you consider merging this PR in that case? and I'll open a new one?

**Review by MarcoGorelli — APPROVED** [review]
Sure, looks good to me - I've finally had a chance to read through the notebook, and I think it's really good!

**Pull request merged**
